// phrontend/lib/index.js
// Prelude
'use strict';

exports.__esModule = true;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _ApiCaller = require('./ApiCaller');

var _ApiCaller2 = _interopRequireDefault(_ApiCaller);

var _ApiError = require('./ApiError');

var _ApiError2 = _interopRequireDefault(_ApiError);

var _AppDispatcher = require('./AppDispatcher');

var _AppDispatcher2 = _interopRequireDefault(_AppDispatcher);

var _NetworkError = require('./NetworkError');

var _NetworkError2 = _interopRequireDefault(_NetworkError);

var _Store = require('./Store');

var _Store2 = _interopRequireDefault(_Store);

if (process.env.NODE_ENV !== 'production') {
  Error.stackTraceLimit = Infinity;
}

exports.ApiCaller = _ApiCaller2['default'];
exports.ApiError = _ApiError2['default'];
exports.Dispatcher = _AppDispatcher2['default'];
exports.NetworkError = _NetworkError2['default'];
exports.Store = _Store2['default'];

// deprecated notice
var deprecatedNotice = 'ampersand-state and ampersand-collection are removed from phrontend\n  Please use v0.0.3 to use ampersand-state and ampersand-collection';

var dummyConstructor = function dummyConstructor() {
  if (process.env.NODE_ENV !== 'production') console.error(deprecatedNotice);
  return {};
};

var State = {
  extend: function extend() {
    return dummyConstructor;
  }
};

exports.State = State;
var Collection = {
  extend: function extend() {
    return dummyConstructor;
  }
};
exports.Collection = Collection;
